<a href="https://colab.research.google.com/github/ajrianop/Estadistica_Descriptiva_y_Python_UD2023-3/blob/main/Parcial1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Parcial 1**

In [ ]:
# Funciones a utilizar

def arithmetic_mean(values : list):
  '''
  Función que permite calcular la media aritmética
  values --> list
  '''
  _sum = 0
  n = len(values)

  for value in values:
      _sum += value  # _sum = _sum + value

  return _sum / n


def geometric_mean(values : list):
  '''
  Función que permite calcular la media geométrica
  values --> list
  '''
  product = 1
  n = len(values)

  for value in values:
      product *= value

  return (product ** (1 / n))

def harmonic_mean(values : list):
  '''
  Función que permite calcular la media armónica
  values --> list
  '''
  _sum = 0
  n = len(values)

  for value in values:
      _sum += 1/value  # Convierte el porcentaje a decimal

  return n/_sum


def quadratic_mean(values : list):
  '''
  Función que permite calcular la media cuadrática
  values --> list
  '''
  _sum = 0
  n = len(values)

  for value in values:
      _sum += value ** 2  # Convierte el porcentaje a decimal

  return (_sum/n)** 0.5


def mid_range(values : list):
  '''
  Función que permite calcular rango medio
  values --> list
  '''
  _min = min(values)
  _max = max(values)

  return ( _min + _max )/2

def median(values : list):
  '''
  Función que permite calcular la mediana
  values --> list
  '''
  n = len(values)
  if n % 2 == 1:
      mediana = values[n // 2]
  else:
      mediana = (values[(n - 1) // 2] + values[n // 2]) / 2
  return mediana

from collections import Counter

def mode(values : list):
  '''
  Función que permite calcular la moda
  values --> list
  '''
  # Se genera la clase counter que tiene un diccionario organizado de mayor a menor por cantidad de votos
  # {key : value}  ---> key : puntuacion ,  value: cantidad de veces repetida la puntuacion
  contador_frecuencia = Counter(values)
  # Generar una lista de tuplas (dato , votación de ese dato)
  modas_resultados = contador_frecuencia.most_common()

  # Obtener los valores de las modas y sus frecuencias
  modas = [moda[0] for moda in modas_resultados if moda[1] == modas_resultados[0][1]]
  frecuencia_maxima = modas_resultados[0][1]
  return modas , frecuencia_maxima

# Medidas de dispersión

def range_func(values : list):
  '''
  Función rango de estadística, la cual consifera punto máximo y mínimo de
  la lista y calcula su diferencia.
  values ---> list
  '''
  return max(values)-min(values)

def mean_deviation(values : list):
  '''
  Función desviación media, la cuál permite ver que tan dispersa esta la
  información con respecto a la media (media aritmética)
  values ---> list
  '''
  _sum = 0
  _mean = arithmetic_mean(values)
  n = len(values)
  for value in values:
    _sum += abs(value - _mean)
  return _sum / n

def variance(datos):
  '''
  Función varianza, la cuál permite ver que tan dispersa esta la
  información con respecto a la media (media aritmética), en este caso con el
  cuadrado de la diferencia.
  values ---> list
  '''
  media = sum(datos) / len(datos)
  varianza = sum((x - media) ** 2 for x in datos) / (len(datos) - 1)
  return varianza

def standard_deviation(values : list):
  '''
  Función desviación estándar, la cuál permite ver que tan dispersa esta la
  información con respecto a la media (media aritmética).
  values ---> list
  '''
  return variance(values) ** (1/2)

def coefficient_of_variation(values : list):
  '''
  Función de coeficientes de variación dados unos valores iniciales.
  Note que el resultado es un porcentaje.
  values ---> list
  '''
  _mean = arithmetic_mean(values)
  _std = standard_deviation(values)

  return (_std  / _mean ) * 100

In [ ]:
# Definimos una función con todas las medidas de tendencia central vistas en clase:

def calculo_med_tend_central(values : list, comentario : str):
  '''
  Función que cálcula:
  Media Aritmética
  Media Geométrica
  Media Armónica
  Media Cuadrática
  Rango Medio
  Mediana
  Moda
  '''
  _dict = dict()
  ari_mean = arithmetic_mean(values)
  geo_mean = geometric_mean(values)
  har_mean = harmonic_mean(values)
  qua_mean = quadratic_mean(values)
  mid_range1 = mid_range(values)
  _median = median(values)
  _mode = mode(values)
  _range_func = range_func(values)
  _mean_deviation = mean_deviation(values)
  _variance = variance(values)
  _standard_deviation = standard_deviation(values)
  coef_var = coefficient_of_variation(values)

  dict_measures = {"arithmetic_mean" :ari_mean , "geometric_mean" : geo_mean, "harmonic_mean" : har_mean, "quadratic_mean" : qua_mean, "mid_range" : mid_range1, "median" : _median , "mode = (mode , frequency)" : _mode ,  "range" :  _range_func , "mean deviation" : _mean_deviation , "variance" : _variance , "standard deviation" : _standard_deviation , "coefficient of variation" : coef_var , "comentario" : comentario}
  for i in dict_measures:
    _dict[i] = dict_measures[i]

  return _dict

## **Ejercicio 1**

Se esta realizando un estudio de resistencia de materiales en vigas de acero para un proyecto de construcción de puentes. Se te proporcionan los siguientes datos de resistencia a la tensión (en MPa) de 8 muestras de acero tomadas de diferentes partes de una viga de puente. Datos: $[450, 470, 480, 485, 490, 500, 510, 520].$

Calcular las siguientes medidas:
1. Media aritmética y media geométrica de los datos.
2. Mediana y moda de los datos.
3. Desviación estándar de los datos. ¿Para que es útil esta medida?

In [ ]:
# Realizamos el cálculo de lista
_list = [450, 470, 480, 485, 490, 500, 510, 520]

_dict = calculo_med_tend_central(_list, 'Cálculo medidas parcial 1, punto 1.')

In [ ]:
import pandas as pd

def dict_to_dataframe(_dict : dict):
  for i in _dict:
    _dict[i] = [_dict[i]]
  df = pd.DataFrame(_dict)
  return df

# Generando la tabla en pandas
dict_to_dataframe(_dict)

arithmetic_mean  geometric_mean  harmonic_mean  quadratic_mean  mid_range  \
0          488.125      487.673457     487.218185      488.572538      485.0   

   median                      mode = (mode , frequency)  range  \
0   487.5  ([450, 470, 480, 485, 490, 500, 510, 520], 1)     70   

   mean deviation    variance  standard deviation  coefficient of variation  \
0          16.875  499.553571           22.350695                  4.578888   

                            comentario  
0  Cálculo medidas parcial 1, punto 1.

1. La **media** es: 488,125. La **media geométrica** es: 487,67.
2. La **mediana** es: 487.5 (note que se tomaron los dos valores centrales y se calculo el promedio ya que teniamos una lista (par) con 8 elementos). La **moda** no es representativa en este aspecto, por lo que todos los valores solo se repiten una sola vez. De este modo se podría decir que la moda es todo el conjunto.
3. La **desviación estándar** es de 22.35, de este modo la información varia de acuerdo a este valor. Es decir que la forma como se dispersan los datos es cercana a las 22.35 unidades.

## **Ejercicio 2**

Se han entrevistado algunas personas de manera incógnita y se han registrado los datos de los salarios anuales que devengan en miles de dólares, salarios : [65, 72, 60, 58, 82, 50, 80, 68, 65, 80, 62, 950, 58, 85, 55, 820, 90, 230, 72, 100]. De este modo una persona considera que \textbf{la media es el mejor valor para considerar un valor central} de toda la población entrevistada. ¿Es adecuado pensar que la mejor medida central para representar la población sea la media? (explique su respuesta, no es necesario hacer cálculos)

**Solución:** Se tiene que dada la información tiene las siguientes medidas centrales y de dispersión:

In [ ]:
# Realizamos el cálculo de lista
_list2 = [65, 72, 60, 58, 82, 50, 80, 68, 65, 80, 62, 950, 58, 85, 55, 820, 90, 230, 72, 100]

_dict2 = calculo_med_tend_central(_list2, 'Cálculo medidas parcial 1, punto 2.')

In [ ]:
import pandas as pd

def dict_to_dataframe(_dict : dict):
  for i in _dict:
    _dict[i] = [_dict[i]]
  df = pd.DataFrame(_dict)
  return df

# Generando la tabla en pandas
dict_to_dataframe(_dict2)

arithmetic_mean  geometric_mean  harmonic_mean  quadratic_mean  mid_range  \
0            160.1       95.133317      78.307286      292.898276      500.0   

   median mode = (mode , frequency)  range  mean deviation      variance  \
0    71.0     ([65, 72, 58, 80], 2)    900          151.97  63323.568421   

   standard deviation  coefficient of variation  \
0          251.641746                157.177855   

                            comentario  
0  Cálculo medidas parcial 1, punto 2.

Dados los salarios de las personas, observamos que hay dos quienes tienen salarios muy altos, por lo cual generaría cambios significativos a la hora de calcular una medida de tendencia central. De esta manera, se plantea que la media (media aritmética) no es un valor muy diciente a la hora de querer determinar un valor central adecuado. Recordamos que para tener una buena descripción de un valor promedio o media, es importante evitar aquellos outliers (valores atípicos).

## **Ejercicio 3**

¿Cual es la estructura de un diccionario en Python? De un ejemplo con pocos datos.

**Solución:** Se tiene que un diccionario está planteado mediante llaves "keys" y valores "values" (que es la información asociada a una llave).

_dict = {"keys" : values}

In [ ]:
_dict = {'nombre1' : [1,2,3,4] , 'nombre2' : (0,9,8,7) , 'nombre3' : "pueden ser string"}
_dict

{'nombre1': [1, 2, 3, 4],
 'nombre2': (0, 9, 8, 7),
 'nombre3': 'pueden ser string'}